### ISSUE TITLE
Losses with Reduction.NONE do not keep the input shape (the result is averaged along the last axis)

### ISSUE THREAD

**Issue by matgad** [tensorflow/tensorflow#27190]
**System information**
- OS Platform and Distribution: **Linux Manjaro**
- TensorFlow installed from: **binary**
- TensorFlow version (use command below): **2.0.0-dev20190326** (nightly build)
- Python version: **3.7.2**

**Describe the current behavior**
Currently, when a loss object is created with reduction=tf.losses.Reduction.NONE, the result is averaged along the last axis. This happens with all the losses in the tf.losses module.

**Describe the expected behavior**
When a loss object is created with reduction=tf.losses.Reduction.NONE, the output shape should match the input shape (y_true and y_pred shape).
This is also reported in the doc: [tf.losses.Reduction](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/losses/Reduction)

Moreover, when a  _sample_weight_ parameter ([BinaryCrossentropy](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/losses/BinaryCrossentropy)) is provided with the same shape of y_pred, it throws an exception. 

**Code to reproduce the issue**

Loss function with Reduction.NONE 
```
bce = tf.losses.BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)
y_true = [1., 1., 1., 1.]
y_pred = [1., 1., 0.5, 0.5]
loss = bce(y_true, y_pred)
print('Loss: ', loss.numpy())
```
CURRENT RESULT: 0.34657347
EXPECTED RESULT: [0., 0., 0.69314694, 0.69314694]

Loss function with Reduction.NONE and sample_weight
```
bce = tf.losses.BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)
y_true = [1., 1., 1., 1.]
y_pred = [1., 1., 0.5, 0.5]
sample_weight = [1, 1, 2, 1]
loss = bce(y_true, y_pred, sample_weight=sample_weight)
print('Loss: ', loss.numpy())
```
CURRENT RESULT: Exception
EXPECTED RESULT: [-0., -0., 1.3862939, 0.69314694]

**Other info / logs**
Currently, to achieve the expected result, an additional "fake" dimension must be added to y_true and y_pred, as in the following example.
```
bce = tf.losses.BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)
y_true = [1., 1., 1., 1.]
y_pred = [1., 1., 0.5, 0.5]
sample_weight = [1, 1, 2, 1]
loss = bce(tf.expand_dims(y_true, -1), tf.expand_dims(y_pred, -1), sample_weight=sample_weight)
print('Loss: ', loss.numpy())
```

**Comment by jvishnuvardhan** [comment]
@amitsrivastava78 Can you please take a look at this issue? Thanks!

**Comment by amitsrivastava78** [comment]
@matgad , actually BinaryCrossentropy uses the function nn.sigmoid_cross_entropy_with_logits which takes into account the reduction and  applies applying the below formula :-

_

> max(x, 0) - x * z + log(1 + exp(-abs(x)))
> where in your case y_pred = x and y_true = z.

_

So based on this the computed size is same as that of input, but there is a twist in this before the actual result is passed it is done a reduce_mean with axis as -1 which means it will add the axis with input dimension and do the mean on tat axes, which s the result you are getting : -
0.34657347

Hope this clears things.

Regards
Amit

**Comment by matgad** [comment]
Hi @amitsrivastava78, thanks for the answer. The behavior that you described is the default one. But when you define a loss with the optional parameter `reduction=tf.losses.Reduction.NONE`, the last reduce_mean (or any other reduce) should be skipped. This issue only applies with None reduction.
`BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)` is just an example, the same behavior should be extended for any loss function.

**Comment by amitsrivastava78** [comment]
@matgad ,i see your point, i think this is a BUG, irrespective of the reduction, the tensorflow is applying the reduce_mean to everything, in the coming days i will try to raise one PR to fix this issue.

Regards
Amit

**Comment by matgad** [comment]
Great, many thanks. Please also take a look on the exception that is thrown when `sample_weight` has the same shape of the input (see the example in my first post for details), I think the two issues are related.

**Comment by amitsrivastava78** [comment]
@matgad , i have raised the PR and included your cases as unit test cases as well, Lets wait for this PR to merge, in the mean time you can  use this PR in case you are blocked.

#27784 

Regards
Amit

**Comment by matgad** [comment]
Wonderful, thank you.

**Comment by pavithrasv** [comment]
@matgad @amitsrivastava78  The loss functions actually do not perform any reduction by default. The expectation is that the input passed to the functions will be atleast 2D. These functions return one loss value per sample as output.

Eg if y_true and y_pred have the shape (3, 3), return value will be of shape (3,). => we get one loss value per sample
if y_true and y_pred have the shape (3, 2, 4), return value will be of shape (3, 2) => we get one loss value per sample, per timestep etc.

Sorry that the functions do not have any documentation about this currently. We are working on adding that, please feel free to contribute to the documentation if you are interested.

I just replied on the PR with this as well.

**Comment by pat-coady** [comment]
Why would issue be closed before updating the documentation? I just ran into the problem because documentation specifically said reduction=Reduction.NONE would give me same shape as I send in.

**Comment by pavithrasv** [comment]
@pat-coady sure we can leave this issue open until we have documentation updated.

**Comment by pat-coady** [comment]
@pavithrasv If it is helpful, I can update docstrings in appropriate places and submit the PR. I haven't contributed to the project before, but am reviewing Contributing Guidelines now.

**Comment by pavithrasv** [comment]
@pat-coady Oh yes absolutely. You are welcome to make contributions anytime :) As mentioned in my comment above we will need to add documentation for all standalone functions in losses.py (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L752) indicating that the input needs to be at least 2D.

**Comment by pat-coady** [comment]
Submitted PR:
https://github.com/tensorflow/tensorflow/pull/30463

**Comment by pavithrasv** [comment]
Great thank you!

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved by the PR #30463 (merged already), but please let me know if I'm mistaken.Thanks!

**Comment by pGit1** [comment]
This still is NOT resolved even with TF2.2. 

```
a = np.array([[.3, .4], [.5, .6]])
b = np.array([[.312, .433], [.555, .66]])

print('incorrect output shape:', tf.keras.losses.BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)(a, b))
print('+'*10)
print('correct output shape:', tf.nn.sigmoid_cross_entropy_with_logits(a, b))

incorrect output shape: tf.Tensor([0.64322317 0.69003338], shape=(2,), dtype=float64)
++++++++++
correct output shape: tf.Tensor(
[[0.76766615 0.75970248]
 [0.73116606 0.6806367 ]], shape=(2, 2), dtype=float64)

```

**Comment by jvishnuvardhan** [comment]
@pGit1 This is expected and this is what @pavithrasv described in detail [here](https://github.com/tensorflow/tensorflow/issues/27190#issuecomment-487233022).

Please check the usage examples and description [here]( https://www.tensorflow.org/api_docs/python/tf/keras/losses/BinaryCrossentropy?version=nightly).

Please let me know if there is anything I am missing. Thanks!

**Comment by pGit1** [comment]
@jvishnuvardhan I see that but it still doesn't make sense to me. I discovered the tf.nn.sigmoid_cross_entropy by accident and it behaves _exactly as one would expect_ when there is no reduction (i.e. element-wise loss scalars). @pavithrasv explanation doesn't makes sense to me and the documentation doesn't either. It's is weird to me that no reduction actually leads to a loss in tensor dimension...I just cannot wrap my head around on what "sample" means in @pavithrasv explanation or the official documentation. 

If I am comparing two 3x3 tensors with no reduction I have **no idea** why I am not getting a 3x3 loss tensor out. I really may be just misunderstanding what "sample" means here. It almost seems like by sample you mean **row** in y_pred and y_true tensors. Its as if you actually do compute an element-wise loss and once this is done you take the mean/sum along the -1 dimension. But the fact that a mean or sum is taken means there is a **reduction** by definition. :(

**Comment by pavithrasv** [comment]
@pGit1 Sorry about that. I understand this can be confusing. This is by design and has been since the beginning of Keras, so any changes to this will be a big breaking change. If your data is 1D, we request that this be converted to 2D as losses expect 2D data. You will get 1D data back ie. one loss value per example when reduction is None.

**Comment by pGit1** [comment]
@pavithrasv I guess adding an extra dimension solution isnt so bad but I still think its weird. Thanks for clarifying.

**Comment by pGit1** [comment]
@pavithrasv @jvishnuvardhan 

Quick question by the way while on this topic. I wonder what tf.keras is actually doing with  tensors in general because if I use a tf.keras.Reshape layer I can literally reshape the layer to anything I want and it will still work. Also I can have an output layer for a model of some fixed dimension and pass it a tensor of different dimension during training and it still works. Why is that?? What is going on now? Im using TF 2.2 by the way. Also tf.reshape behaves as it should, i.e. it breaks when I try to arbitrarily reshape a tensor with more or less elements than input tensor...

**Comment by BioWar** [comment]
> 
> 
> Hi @amitsrivastava78, thanks for the answer. The behavior that you described is the default one. But when you define a loss with the optional parameter `reduction=tf.losses.Reduction.NONE`, the last reduce_mean (or any other reduce) should be skipped. This issue only applies with None reduction.
> `BinaryCrossentropy(reduction=tf.losses.Reduction.NONE)` is just an example, the same behavior should be extended for any loss function.

That was exactly the case in my custom loss function created in class inherited from **tf.keras.losses.Loss** . After deletion of all **reduce_mean**/**reduce_sum** the function works like a charm. Thank you so much for your short explanation!!!

**Comment by llodds** [comment]
Hi, I was reading the distributed strategy and found one issue in the official doc which is related to this thread.

https://www.tensorflow.org/tutorials/distribute/custom_training

> If labels is multi-dimensional, then average the per_example_loss across the number of elements in each sample. For example, if the shape of predictions is (batch_size, H, W, n_classes) and labels is (batch_size, H, W), you will need to update per_example_loss like: per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:]), tf.float32)

Since the last dimension is automatically reduced in "NONE" mode, then the manual reduce formula for the distributed.strategy should be
per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:**-1**]), tf.float32) # notice the -1 in shape() to remove the accumulation for the last axis

I don't know which place I should put this post to, but please update the relevant people who is in charge of the doc. Thanks.

**Comment by chethanjjj** [comment]
If we set the reduction in our loss totf.losses.Reduction.NONE when we compile our model is there a way to retrieve the unreduced losses when we evaluate on a sample? I'm running into an issue where the loss in my model (with zeroed-out weights and bias - code below on how i'm doing this) is not what I'm expecting. 

```
    for ix, layer in enumerate(model.layers):
      weights = layer.get_weights()
      for arr in weights:
        arr[(arr > 0) | (arr < 0)] = 0
      layer.set_weights(weights)
```


So I'd like to diagnose how my model is calculating the loss by looking at output from the 3 stages of reduction: NONE, SUM, and SUM_OVER_BATCH_SIZE

**Comment by chengstark** [comment]
Issue still exists on CoLab
